internal/metrics/metrics.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/ChainSafe/gossamer/internal/httpserver"
+	"github.com/ChainSafe/gossamer/internal/log"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const defaultInterval = 10 * time.Second
+
+var logger log.LeveledLogger = log.NewFromGlobal(log.AddContext("pkg", "metrics"))
+
+// IntervalConfig for interval collection
+type IntervalConfig struct {
+	Publish  bool
+	Interval time.Duration
+}
+
+// NewIntervalConfig is constructor for IntervalConfig, and uses default metrics interval
+func NewIntervalConfig(publish bool) IntervalConfig {
+	return IntervalConfig{
+		Publish:  publish,
+		Interval: defaultInterval,
+	}
+}
+
+// Server is a metrics http server
+type Server struct {
+	cancel context.CancelFunc
+	server *httpserver.Server
+	done   chan error
+}
+
+// NewServer is a constructor for metrics server
+func NewServer(address string) (s *Server) {
+	m := http.NewServeMux()
+	m.Handle("/metrics", promhttp.Handler())
+	s = &Server{
+		server: httpserver.New("metrics", address, m, logger),
+	}
+	return
+}
+
+// Start will start a dedicated metrics server at the given address.
+func (s *Server) Start(address string) (err error) {
+	logger.Infof("Starting metrics server at http://%s/metrics", address)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	s.cancel = cancel
+	ready := make(chan struct{})
+	s.done = make(chan error)
+
+	go s.server.Run(ctx, ready, s.done)
+
+	select {
+	case <-ready:
+		return nil
+	case err := <-s.done:
+		close(s.done)
+		if err != nil {
+			return err
+		}
+		return fmt.Errorf("metrics server exited unexpectedly")
+	}
+}
+
+// Stop will stop the metrics server
+func (s *Server) Stop() (err error) {
+	s.cancel()
+	select {
+	case err := <-s.done:
+		close(s.done)
+		if err != nil {
+			return err
+		}
+		return fmt.Errorf("metrics server exited unexpectedly")
+	case <-time.NewTimer(30 * time.Second).C:
```

Comment:
declare this before the `select` block and drain in the `case err := <-done` case with

```go
if !timer.Stop() {
	<-timer.C
}
```

Or at the very least stop the timer in that first case, the channel race is unlikely given the 30 wait time.